src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test/ProjectConfigurationStateSynchronizerTest.cs

```diff
@@ -446,30 +459,43 @@ public async Task ProjectConfigurationFileChanged_Changed_CantDeserialize_Resets
     }
 
     [Fact]
-    public async Task ProjectConfigurationFileChanged_Changed_UntrackedProject_Noops()
+    public async Task ProjectConfigurationFileChanged_Changed_UntrackedProject_CallsUpdate()
     {
         // Arrange
-        var projectService = new StrictMock<IRazorProjectService>();
-
-        using var synchronizer = GetSynchronizer(projectService.Object);
-        var synchronizerAccessor = synchronizer.GetTestAccessor();
-
         var changedArgs = new ProjectConfigurationFileChangeEventArgs(
             configurationFilePath: "/path/to/project.razor.bin",
             kind: RazorFileChangeKind.Changed,
             deserializer: CreateDeserializer(projectInfo: null));
 
+        var intermediateOutputPath = FilePathNormalizer.GetNormalizedDirectoryName(changedArgs.ConfigurationFilePath);
+        var projectKey = TestProjectKey.Create(intermediateOutputPath);
+
+        var projectServiceMock = new StrictMock<IRazorProjectService>();
+        projectServiceMock
+            .Setup(p => p.UpdateProjectAsync(
+                projectKey,
+                It.IsAny<RazorConfiguration>(),
+                It.IsAny<string>(),
+                It.IsAny<string>(),
+                It.IsAny<ProjectWorkspaceState>(),
+                It.IsAny<ImmutableArray<DocumentSnapshotHandle>>(),
+                It.IsAny<CancellationToken>()))
+            .Returns(Task.CompletedTask);
+
+        using var synchronizer = GetSynchronizer(projectServiceMock.Object);
+        var synchronizerAccessor = synchronizer.GetTestAccessor();
+
         // Act
         synchronizer.ProjectConfigurationFileChanged(changedArgs);
 
         await synchronizerAccessor.WaitUntilCurrentBatchCompletesAsync();
 
         // Assert
-        projectService.VerifyAll();
+        projectServiceMock.VerifyAll();
     }
 
     [Fact]
-    public async Task ProjectConfigurationFileChanged_RemoveThenAdd_OnlyAdds()
+    public async Task ProjectConfigurationFileChanged_RemoveThenAdd_Updates()
```

Comment:
This test didn't do what its name implied it did. It actually verified that if you Add then Change, you get an Add then Update.

The test now does what its name implies, which is actually testing the restored functionality of collapsing removes and adds into a single Update operation.

Reply:
Thank you! These tests made my eyes cross. :smile: